src/altair/block_processing.rs

```diff
@@ -29,9 +46,123 @@ pub fn process_attestation<
+            InvalidAttestation::InvalidTargetEpoch {
+                target: data.target.epoch,
+                current: current_epoch,
+            },
+        )));
+    }
+
+    let attestation_epoch = compute_epoch_at_slot(data.slot, context);
+    if data.target.epoch != attestation_epoch {
+        return Err(invalid_operation_error(InvalidOperation::Attestation(
+            InvalidAttestation::InvalidSlot {
+                slot: data.slot,
+                epoch: attestation_epoch,
+                target: data.target.epoch,
+            },
+        )));
+    }
+
+    let attestation_has_delay = data.slot + context.min_attestation_inclusion_delay <= state.slot;
+    let attestation_is_recent = state.slot <= data.slot + context.slots_per_epoch;
+    let attestation_is_timely = attestation_has_delay && attestation_is_recent;
+    if !attestation_is_timely {
+        return Err(invalid_operation_error(InvalidOperation::Attestation(
+            InvalidAttestation::NotTimely {
+                state_slot: state.slot,
+                attestation_slot: data.slot,
+                lower_bound: data.slot + context.slots_per_epoch,
+                upper_bound: data.slot + context.min_attestation_inclusion_delay,
+            },
+        )));
+    }
+
+    let committee_count = get_committee_count_per_slot(state, data.target.epoch, context);
+    if data.index >= committee_count {
+        return Err(invalid_operation_error(InvalidOperation::Attestation(
+            InvalidAttestation::InvalidIndex {
+                index: data.index,
+                upper_bound: committee_count,
+            },
+        )));
+    }
+
+    let committee = get_beacon_committee(state, data.slot, data.index, context)?;
+    if attestation.aggregation_bits.len() != committee.len() {
+        return Err(invalid_operation_error(InvalidOperation::Attestation(
+            InvalidAttestation::Bitfield {
+                expected_length: committee.len(),
+                length: attestation.aggregation_bits.len(),
+            },
+        )));
+    }
+
+    // Participation flag indices
+    let inclusion_delay = state.slot - data.slot;
+    let participation_flag_indices =
+        get_attestation_participation_flag_indices(state, data, inclusion_delay, context)?;
+
+    // Verify signature
+    let _ = is_valid_indexed_attestation(
+        state,
+        &mut get_indexed_attestation(state, attestation, context)?,
+        context,
+    )?;
+
+    // Update epoch participation flags
+    // @dev deviate from the order of the spec to avoid immutable borrow after mutable borrow
+    let attesting_indices =
+        get_attesting_indices(state, data, &attestation.aggregation_bits, context)?;
+
+    let epoch_participation = if is_current {
+        &mut state.current_epoch_participation
+    } else {
+        &mut state.previous_epoch_participation
+    };
+
+    let mut proposer_reward_numerator = 0;
+    // @dev this is where I start having trouble
```

Comment:
I at least got something to work using a `HashMap`, similar to what was recommended in `process_sync_aggregate`. Can push once these other cleanups are confirmed...but here's what I've done:
https://gist.github.com/dtbuchholz/9ca416cda0d4b368d873bedee8c16eef

Reply:
yeah this seems like it would work!

I think we may be able to coerce the borrow checker some and get something a bit simpler, let me try something

Reply:
so this is a bit annoying but I would just do something like this:

https://gist.github.com/rust-play/149e04c2a657723f4d3fb477208881e3

the borrow checker doesn't like that we have a mutable borrow into `state` across the span of the loop where we also pass a second mutable borrow of the entire thing

if we had mutable borrows of disjoint fields of `state` this would be fine but I don't think it is worth rewriting `get_base_reward` at this time

Reply:
ahh yeah, much simpler